python/lsst/ip/diffim/registerImage.py

```diff
@@ -0,0 +1,192 @@
+        """Fit Wcs to matches
+
+        The fitting includes iterative sigma-clipping.
+
+        @param matches: List of matches (first is target, second is input)
+        @param inputWcs: Original input Wcs
+        @param inputBBox: Bounding box of input image
+        @return Wcs
+        """
+        copyMatches = type(matches)(matches)
+        refCoordKey = copyMatches[0].first.getTable().getCoordKey()
+        inCentroidKey = copyMatches[0].second.getTable().getCentroidKey()
+        for i in range(self.config.sipIter):
+            sipFit = makeCreateWcsWithSip(copyMatches, inputWcs, self.config.sipOrder, inputBBox)
+            self.log.debug("Registration WCS RMS iteration %d: %f pixels",
+                           i, sipFit.getScatterInPixels())
+            wcs = sipFit.getNewWcs()
+            dr = [m.first.get(refCoordKey).angularSeparation(
+                wcs.pixelToSky(m.second.get(inCentroidKey))).asArcseconds() for
+                m in copyMatches]
+            dr = numpy.array(dr)
+            rms = math.sqrt((dr*dr).mean())  # RMS from zero
+            rms = max(rms, 1.0e-9)  # Don't believe any RMS smaller than this
+            self.log.debug("Registration iteration %d: rms=%f", i, rms)
+            good = numpy.where(dr < self.config.sipRej*rms)[0]
+            numBad = len(copyMatches) - len(good)
+            self.log.debug("Registration iteration %d: rejected %d", i, numBad)
+            if numBad == 0:
+                break
+            copyMatches = type(matches)(copyMatches[i] for i in good)
+
+        sipFit = makeCreateWcsWithSip(copyMatches, inputWcs, self.config.sipOrder, inputBBox)
+        self.log.info("Registration WCS: final WCS RMS=%f pixels from %d matches" %
+                      (sipFit.getScatterInPixels(), len(copyMatches)))
+        self.metadata.set("SIP_RMS", sipFit.getScatterInPixels())
+        self.metadata.set("SIP_GOOD", len(copyMatches))
+        self.metadata.set("SIP_REJECTED", len(matches) - len(copyMatches))
+        wcs = sipFit.getNewWcs()
+        return wcs
+
+    def warpExposure(self, inputExp, newWcs, templateWcs, templateBBox):
+        """Warp input exposure to template frame
+
+        There are a variety of data attached to the exposure (e.g., PSF, Calib
+        and other metadata), but we do not attempt to warp these to the template
+        frame.
+
+        @param inputExp: Input exposure, to be warped
+        @param newWcs: Revised Wcs for input exposure
+        @param templateWcs: Target Wcs
+        @param templateBBox: Target bounding box
+        @return Warped exposure
+        """
+        warper = Warper.fromConfig(self.config.warper)
+        copyExp = inputExp.Factory(inputExp.getMaskedImage(), newWcs)
+        alignedExp = warper.warpExposure(templateWcs, copyExp, destBBox=templateBBox)
+        return alignedExp
+
+    def warpSources(self, inputSources, newWcs, templateWcs, templateBBox):
+        """Warp sources to the new frame
```

Comment:
You should note that only points that fall inside templateBBox are included in the output sources. It took a bit of staring at the code to realize that.